kable-core/src/jvmMain/kotlin/com/juul/kable/btleplug/BtleplugPeripheral.kt

```diff
@@ -0,0 +1,218 @@
+        if (!ffi.connect(cancellationHandle)) {
+            throw IOException("Failed to connect.")
+        }
+        suspendUntil<State.Connecting.Services>()
+
+        logger.info { message = "Discovering services" }
+        ffi.discoverServices()
+        _services.value = ffi.services().map(::BtleplugService)
+        ServicesDiscoveredPeripheral(this).run { onServicesDiscovered() }
+
+        logger.verbose { message = "Configuring characteristic observations" }
+        _state.value = State.Connecting.Observes
+        observers.onConnected()
+        _state.value = State.Connected(scope)
+
+        return scope
+    }
+
+    override suspend fun connect(): CoroutineScope =
+        connectAction.awaitConnect()
+
+    override suspend fun disconnect() {
+        logger.verbose { message = "Disconnect request" }
+        _state.value = State.Disconnecting
+        connectAction.cancelAndJoin(CancellationException(NotConnectedException("Disconnect requested")))
+        _state.value = Disconnected()
+    }
+
+    // Btleplug doesn't support retrieving the MTU, so we're stuck on the default
+    override suspend fun maximumWriteValueLengthForType(writeType: WriteType): Int = 20
+
+    @ExperimentalApi
+    override suspend fun rssi(): Int = withContext(Dispatchers.IO) {
+        ffi.properties().rssi?.toInt() ?: Int.MIN_VALUE
+    }
+
+    override suspend fun read(characteristic: Characteristic): ByteArray {
+        logger.verbose { message = "Reading from $characteristic" }
+        return withContext(Dispatchers.IO) {
+            ffi.read(getCharacteristic(characteristic))
+        }
+    }
+
+    override suspend fun read(descriptor: Descriptor): ByteArray {
+        logger.verbose { message = "Reading from $descriptor" }
+        return withContext(Dispatchers.IO) {
+            ffi.readDescriptor(
+                getDescriptor(
+                    descriptor.serviceUuid.toString(),
+                    descriptor.characteristicUuid.toString(),
+                    descriptor.descriptorUuid.toString(),
+                ),
+            )
+        }
+    }
+
+    override suspend fun write(characteristic: Characteristic, data: ByteArray, writeType: WriteType) {
+        logger.verbose { message = "Writing to $characteristic, type=$writeType data=${data.size} bytes" }
+        return withContext(Dispatchers.IO) {
+            ffi.write(getCharacteristic(characteristic), data, writeType.ffi())
+        }
+    }
+
+    override suspend fun write(descriptor: Descriptor, data: ByteArray) {
+        logger.verbose { message = "Writing to $descriptor, data=${data.size} bytes" }
+        return withContext(Dispatchers.IO) {
+            ffi.writeDescriptor(
+                getDescriptor(
+                    descriptor.serviceUuid.toString(),
+                    descriptor.characteristicUuid.toString(),
+                    descriptor.descriptorUuid.toString(),
+                ),
+                data,
+            )
+        }
```

Comment:
How does this operation handle cancellation (i.e. when caller, on the Kotlin side, cancels the coroutine)?
Does the rust side honor the cancellation? Does it even get the signal?